src/os/bluestore/BlueStore.cc

```diff
@@ -9015,7 +9015,210 @@ int BlueStore::_split_collection(TransContext *txc,
   return r;
 }
 
+#define KEY_SLAB 32
+#define VALUE_SLAB 64
 
+int get_key_slab(size_t sz)
+{
+  return (sz/KEY_SLAB);
+}
+
+string get_key_slab_to_range(int slab)
+{
+  int lower_bound = slab * KEY_SLAB;
+  int upper_bound = (slab + 1) * KEY_SLAB;
+  string ret = "[" + stringify(lower_bound) + "," + stringify(upper_bound) + ")";
+  return ret;
+}
+
+int get_value_slab(size_t sz)
+{
+  return (sz/VALUE_SLAB);
+}
+
+string get_value_slab_to_range(int slab)
+{
+  int lower_bound = slab * VALUE_SLAB;
+  int upper_bound = (slab + 1) * VALUE_SLAB;
+  string ret = "[" + stringify(lower_bound) + "," + stringify(upper_bound) + ")";
+  return ret;
+}
+
+//Itrerates through the db and collects the stats
+void BlueStore::get_db_histogram(Formatter *f)
+{
+  //globals
+  uint64_t num_onodes = 0;
+  uint64_t num_shards = 0;
+  uint64_t num_super = 0;
+  uint64_t num_coll = 0;
+  uint64_t num_omap = 0;
+  uint64_t num_wal = 0;
+  uint64_t num_alloc = 0;
+  uint64_t num_stat = 0;
+  uint64_t num_others = 0;
+  uint64_t num_shared_shards = 0;
+  size_t max_key_size =0, max_value_size = 0;
+  uint64_t total_key_size = 0, total_value_size = 0;
+  size_t key_size = 0, value_size = 0;
+  int key_slab, value_slab;
+
+  //histogram
+  struct value_dist
+  {
+    uint64_t count;
+    uint32_t max_len;
+  };
+  typedef struct value_dist value_dist_t;
+
+  struct key_dist
+  {
+    uint64_t count;
+    uint32_t max_len;
+    map<int, value_dist_t> val_map; // slab id to count and others
+  };
+  typedef struct key_dist key_dist_t;
 
+  typedef map<int, key_dist_t> key_map_t; // contains the key slab and count
+  map<string, key_map_t > key_hist;
+  map<int, uint64_t> value_hist; // only value distribution histogram
 
+
+  utime_t start = ceph_clock_now(NULL);
+
+  KeyValueDB::WholeSpaceIterator iter = db->get_iterator();
+  iter->seek_to_first();
+  while (iter->valid()) {
+    dout(30) << __func__ << " Key: " << iter->key() << dendl;
+    key_size = iter->key_size();
+    value_size = iter->value_size();
+    key_slab = get_key_slab(key_size);
+    value_slab = get_value_slab(value_size);
+    value_hist[value_slab]++;
+    max_key_size = MAX(max_key_size, key_size);
+    max_value_size = MAX(max_value_size, value_size);
+    total_key_size += key_size;
+    total_value_size += value_size;
+
+
+    pair<string,string> key(iter->raw_key());
+
+    if (key.first == PREFIX_SUPER) {
+	key_hist[PREFIX_SUPER][key_slab].count++;
```

Comment:
Suggest to introduce a helper func  to update a specific key_hist entry in a single call.
e.g.
update_entry( key_hist, PREFIX_SUPER, key_size, value_size);